### PR TITLE
VPI: Add support for variables of real and string data types

### PR DESCRIPTION
**Pull request by marlonjames**
I updated the test, but I am not sure how to add the support for the variable types.

### CODE REVIEW

File: test_regress/t/t_vpi_var.cpp [hunk 1]
```diff
@@ -88,6 +90,13 @@ bool verbose = false;
         return __LINE__; \
     }
 
+#define CHECK_RESULT_REAL(got, exp, delta) \
```
Comment (wsnyder) [inline test_regress/t/t_vpi_var.cpp:93]: Please make a TEST_CHECK_REAL_EQ in TestCheck.h.

File: test_regress/t/t_vpi_var.cpp [hunk 2]
```diff
@@ -11,6 +11,7 @@
 
 #ifdef IS_VPI
 
+#include "sv_vpi_user.h"
```
Comment (wsnyder) [inline test_regress/t/t_vpi_var.cpp:14]: Can then remove the next line

File: include/verilated_vpi.cpp [hunk 3]
```diff
@@ -1749,6 +1749,11 @@ bool vl_check_format(const VerilatedVar* varp, const p_vpi_value valuep, const c
         case VLVT_UINT32: return status;
         default: status = false;
         }
+    } else if (valuep->format == vpiRealVal) {
+        switch (varp->vltype()) {
+        case VLVT_UINT64: return status;
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:1754]:
```suggestion
        case VLVT_REAL: return status;
```

Then you need plumb backwards to add VLVT_REAL, which will reach into the Verilator sources.

**Review by wsnyder — COMMENTED** [review]
Great start.  Mainly for next steps follow how one of the supported types works by grepping the code, and add appropriate cases to handle the newly supported types. E.g. you'd use isReal() in verilator to make a new VLVT_REAL type.

**Comment by marlonjames** [comment]
I wasn't sure the best approach to distinguishing the integer atom types, but that is less important to me than supporting `vpiRealVar`/`vpiStringVar`.
I can take them out of the test, or if you have a suggestion on adding support I can work on that.

File: include/verilated_vpi.cpp [hunk 4]
```diff
@@ -2171,12 +2400,17 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
...
+                *(reinterpret_cast<std::string*>(vop->varDatap())) = valuep->value.str;
+                return object;
+            } else {
+                const int bytes = VL_BYTES_I(vop->varp()->packed().elements());
+                const int len = std::strlen(valuep->value.str);
+                CData* const datap = (reinterpret_cast<CData*>(vop->varDatap()));
+                for (int i = 0; i < bytes; ++i) {
+                    // prepend with 0 values before placing string the least significant bytes
+                    datap[i] = (i < len) ? valuep->value.str[len - i - 1] : 0;
+                }
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:2413]: Suggest do a memset of zero, followed by a memcpy. Those functions will optimize to do 64 bit copies, so will end up faster.

File: include/verilated_vpi.h [hunk 5]
```diff
@@ -30,9 +30,9 @@
 #include "verilated_syms.h"
 
 //======================================================================
-// From IEEE 1800-2009 annex K
+// From IEEE 1800-2017 annex M
```
Comment (wsnyder) [inline include/verilated_vpi.h:33]: Thanks, I just want to make sure you looked at the IEEE 2017 and made sure there were no other changes?

**Review by wsnyder — COMMENTED** [review]
I'd suggest remove the integer size if you don't need it, as I suspect they are unlikely to be needed soon/ever.

**Comment by wsnyder** [comment]
I think it would be cleaner to have the sv_vpi_user added, along with vpi_user update in a separate pre-pull, it's ok if there's no .cpp files or testing (since it's all #defines)., would you mind making that?

File: include/verilated_vpi.h (hunk 5 above)
Comment (marlonjames) [reply include/verilated_vpi.h:33]: From what I can tell you already had `vpi_user.h` from the 2017 standard. There are no changes there.

File: include/verilated_vpi.cpp (hunk 4 above)
Comment (marlonjames) [reply include/verilated_vpi.cpp:2413]: I don't think that will work in this case. The source data is in reverse order.

**Comment by marlonjames** [comment]
I believe this is ready. The code analysis is flagging the `VPI_HANDLE()` macro.

**Pull request merged**